File: src/main/java/kurenai/mybot/handler/Handler.java
package kurenai.mybot.handler;

import kurenai.mybot.QQBotClient;
import kurenai.mybot.TelegramBotClient;
import net.mamoe.mirai.event.events.GroupAwareMessageEvent;
import org.telegram.telegrambots.meta.api.objects.Message;
import org.telegram.telegrambots.meta.api.objects.Update;

public interface Handler {


    default boolean preHandle(TelegramBotClient client, Update update) {
        return true;
    }

    /**
     * @param client
     * @param qqClient
     * @param update
     * @return true 为继续执行，false中断
     */
    default boolean handleMessage(TelegramBotClient client, QQBotClient qqClient, Update update, Message message) throws Exception {
        return true;
    }

    default boolean handleEditMessage(TelegramBotClient client, QQBotClient qqClient, Update update, Message message) throws Exception {
        return true;
    }

    default boolean postHandle(TelegramBotClient client, Update update, Message message) {
        return true;
    }

    default boolean handle(QQBotClient client, TelegramBotClient telegramBotClient, GroupAwareMessageEvent event) throws Exception {
        return true;
    }

    default int order() {
        return 100;
    }

    default String getHandlerName() {
        return this.getClass().getSimpleName();
    }

}

File: src/main/java/kurenai/mybot/handler/Handler.kt
package kurenai.mybot.handler

import kurenai.mybot.QQBotClient
import kurenai.mybot.TelegramBotClient
import net.mamoe.mirai.event.events.GroupAwareMessageEvent
import net.mamoe.mirai.event.events.MessageRecallEvent
import org.telegram.telegrambots.meta.api.objects.Message
import org.telegram.telegrambots.meta.api.objects.Update
import org.telegram.telegrambots.meta.exceptions.TelegramApiException

interface Handler : Comparable<Handler> {
    fun preHandle(client: TelegramBotClient?, update: Update?): Boolean {
        return true
    }

    /**
     * @param client
     * @param qqClient
     * @param update
     * @return true 为继续执行，false中断
     */
    @Throws(Exception::class)
    fun handleMessage(client: TelegramBotClient, qqClient: QQBotClient, update: Update?, message: Message): Boolean {
        return true
    }

    @Throws(Exception::class)
    fun handleEditMessage(client: TelegramBotClient, qqClient: QQBotClient, update: Update?, message: Message): Boolean {
        return true
    }

    fun postHandle(client: TelegramBotClient?, update: Update?, message: Message?): Boolean {
        return true
    }

    @Throws(Exception::class)
    fun handle(client: QQBotClient?, telegramBotClient: TelegramBotClient, event: GroupAwareMessageEvent): Boolean {
        return true
    }

    @Throws(TelegramApiException::class)
    fun handleRecall(client: QQBotClient?, telegramBotClient: TelegramBotClient, event: MessageRecallEvent): Boolean {
        return true
    }

    fun order(): Int {
        return 100
    }

    override fun compareTo(other: Handler): Int {
        return this.order() - other.order()
    }

    fun handleName(): String {
        return this.javaClass.simpleName
    }

}